packages/hardhat-core/src/internal/hardhat-network/stack-traces/vm-debug-tracer.ts

```diff
@@ -319,9 +323,7 @@ export class VMDebugTracer {
       .toString("hex")
       .match(/.{1,64}/g);
 
-    const result = memory === null ? [] : [...memory];
-
-    return result;
+    return memory === null ? [] : memory.slice(0, Number(step.memoryWordCount)); // Remove the additional non allocated memory
```

Comment:
I can't add a code suggestion for the whole block, but what do you think about this? I think its more readable

```js
    const rawMemory =
      Buffer.from(step.memory)
        .toString("hex")
        .match(/.{1,64}/g) ?? [];

    // Remove the additional non allocated memory
    return rawMemory.slice(0, Number(step.memoryWordCount));
```